### PR TITLE
Rewrite of process engine to support actions in tasks resolves #205 and #207

### LINKED ISSUE #205: Endpoint providing information what actions a user has access to
### Description

The app backend needs to provide an endpoint providing information about the requesting users access rights on the current step.

This should cover:
* read
* write
* actions
  * sign
  * reject
  * confirm
  * instantiate

`confirm` and `instantiate` are not for this issue, but is included to show that we need to support multiple actions related to a step in the process and that we can control access to these actions.

The same endpoint should also list the roles that actually has access to the current step and actions.

### Additional Information
Related epic: https://github.com/Altinn/app-template-dotnet/issues/170
Related frontend issue: https://github.com/Altinn/app-frontend-react/issues/991

### Tasks

_No response_

### Proposed change in process file and process engine
Given this fairly simple, and not uncommon flow: 
image

Implicit actions that happen when an instance leave a Data task is:
1. Lock the data associated with the task
2. Generate PDF for the task

Moving "back" in the process we need to delete the existing PDF to ensure it has no outdated data and unlock the data to make changes possible.

To handle the unlock and delete actions we need some way of knowing if the process is moving forwards or backwards 

#### Today
Todays solution of informing the process engine if the process is moving forward or backwards is done by adding custom attributes to the flows out of the gateway. Example BPMN describing the image above (note the altinn:flowtype on SequenceFlow 4 and 5):
```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn2:definitions id="Altinn_Data_Confirmation_Process_Definition"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xmlns:bpmn2="http://www.omg.org/spec/BPMN/20100524/MODEL"
  xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI"
  xmlns:dc="http://www.omg.org/spec/DD/20100524/DC"
  xmlns:di="http://www.omg.org/spec/DD/20100524/DI"
  xmlns:altinn="http://altinn.no"
  xsi:schemaLocation="http://www.omg.org/spec/BPMN/20100524/MODEL BPMN20.xsd"
  targetNamespace="http://bpmn.io/schema/bpmn" >
  <bpmn2:process id="Altinn_Data_Confirmation_Process" isExecutable="false">
    <bpmn2:startEvent id="Start">
      <bpmn2:outgoing>SequenceFlow_1</bpmn2:outgoing>
    </bpmn2:startEvent>
    <bpmn2:task id="Task1" name="Task1:Data" altinn:tasktype="data">
      <bpmn2:incoming>SequenceFlow_1</bpmn2:incoming>
      <bpmn2:incoming>SequenceFlow_5</bpmn2:incoming>
      <bpmn2:outgoing>SequenceFlow_2</bpmn2:outgoing>
    </bpmn2:task>
    <bpmn2:task id="Task2" name="Task2:Confirm" altinn:tasktype="confirmation">
      <bpmn2:incoming>SequenceFlow_2</bpmn2:incoming>
      <bpmn2:outgoing>SequenceFlow_3</bpmn2:outgoing>
    </bpmn2:task>
	<bpmn2:exclusiveGateway id="Gateway_1">
	    <bpmn2:incoming>SequenceFlow_3</bpmn2:incoming>
	    <bpmn2:outgoing>SequenceFlow_4</bpmn2:outgoing>
	    <bpmn2:outgoing>SequenceFlow_5</bpmn2:outgoing>
	</bpmn2:exclusiveGateway>
    <bpmn2:endEvent id="End">
      <bpmn2:incoming>SequenceFlow_4</bpmn2:incoming>
    </bpmn2:endEvent>
    <bpmn2:sequenceFlow id="SequenceFlow_1" sourceRef="Start" targetRef="Task1" />
    <bpmn2:sequenceFlow id="SequenceFlow_2" sourceRef="Task1" targetRef="Task2" />
    <bpmn2:sequenceFlow id="SequenceFlow_3" sourceRef="Task2" targetRef="Gateway_1" />
	<bpmn2:sequenceFlow id="SequenceFlow_4" sourceRef="Gateway_1" targetRef="End" altinn:flowtype="CompleteCurrentMoveToNext" />
	<bpmn2:sequenceFlow id="SequenceFlow_5" sourceRef="Gateway_1" targetRef="Task1" altinn:flowtype="AbandonCurrentReturnToNext" />
  </bpmn2:process>
</bpmn2:definitions>
```
To make present the user with the available choices, lets call them "actions", that is possible from this task the process engine needs to do a simulated play of the process and present the user with all the tasks that could be reached by flows from the current task, gateways are evaluated and if not decision can be made all flows are followed and resulting tasks presented as possible "actions" for the user.

In the example above, if the process is in Task2 the user would be presented with a list of possible actions: [Task1, End].
If the user chooses to take the "action" Task1 the process engine would follow the flows from Task2 to Task1 and discover that SequenceFlow_5 has flowtype AbandonCurrentReturnToNext and execute logic connected to this flowtype.

In my opinion this is a flawed, if not wrong, when the process is on a given task I think this Task should contain enough information to present the user with what "actions" are possible to do without reading the rest of the process definition.

#### Proposal
TL;DR; 
* Remove altinn:flowtype attribute and introduce Extensions properties on Tasks that defines available actions.
* Remove altinn:tasktype attribute and move it to the extension properties to adhere to the bpmn specification.

Frontend would then get the possible actions from the instance current position(task) in the process. No need to follow flows or evaluating gateways before the user actually takes a action.
Frontend still receives the altinn:tasktype but it is derived from the extension properties instead
Unlocking the data and deleting outdated PDFs would then be performed before the process leaves Task2 in the example process.

Example process definition:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL"
  xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI"
  xmlns:dc="http://www.omg.org/spec/DD/20100524/DC"
  xmlns:altinn="http://altinncdn.no/schemas/1.0/bpmn"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xmlns:di="http://www.omg.org/spec/DD/20100524/DI"
  xmlns:modeler="http://camunda.org/schema/modeler/1.0" id="Definitions_1exv6jf"
  targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.5.0"
  modeler:executionPlatform="Camunda Platform" modeler:executionPlatformVersion="7.18.0">
  <bpmn:process id="Process_1om6n79" isExecutable="true">
    <bpmn:startEvent id="Start">
      <bpmn:outgoing>SequenceFlow_1</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:task id="Task1" name="Task1:Data">
      <bpmn:extensionElements>
        <altinn:properties>
          <altinn:actions>
            <altinn:action id="submit" name="Send inn" />
          </altinn:actions>
          <altinn:tasktype id="data" />
        </altinn:properties>
      </bpmn:extensionElements>
      <bpmn:incoming>SequenceFlow_1</bpmn:incoming>
      <bpmn:incoming>SequenceFlow_5</bpmn:incoming>
      <bpmn:outgoing>SequenceFlow_2</bpmn:outgoing>
    </bpmn:task>
    <bpmn:task id="Task2" name="Task2:Confirm">
      <bpmn:extensionElements>
        <altinn:properties>
          <altinn:actions>
            <altinn:action id="confirm" name="Bekreft" />
            <altinn:action id="reject" name="Avslå" />
          </altinn:actions>
          <altinn:tasktype id="confirm" />
        </altinn:properties>
      </bpmn:extensionElements>
      <bpmn:incoming>SequenceFlow_2</bpmn:incoming>
      <bpmn:outgoing>SequenceFlow_3</bpmn:outgoing>
    </bpmn:task>
    <bpmn:exclusiveGateway id="Gateway_1">
      <bpmn:incoming>SequenceFlow_3</bpmn:incoming>
      <bpmn:outgoing>SequenceFlow_4</bpmn:outgoing>
      <bpmn:outgoing>SequenceFlow_5</bpmn:outgoing>
    </bpmn:exclusiveGateway>
    <bpmn:endEvent id="End">
      <bpmn:incoming>SequenceFlow_4</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id="SequenceFlow_1" sourceRef="Start" targetRef="Task1" />
    <bpmn:sequenceFlow id="SequenceFlow_2" sourceRef="Task1" targetRef="Task2" />
    <bpmn:sequenceFlow id="SequenceFlow_3" sourceRef="Task2" targetRef="Gateway_1" />
    <bpmn:sequenceFlow id="SequenceFlow_4" sourceRef="Gateway_1" targetRef="End" />
    <bpmn:sequenceFlow id="SequenceFlow_5" sourceRef="Gateway_1" targetRef="Task1" />
  </bpmn:process>
</bpmn:definitions>
```

#### API response proposed by frontend
- `PUT */process/next` (response)
- `GET */process` (response)
```json5
{
  ...
  "currentTask": {
    ...
    "read": boolean,
    "write": boolean,
    "actions": {
      "instantiate": boolean,
      "confirm": boolean,
      "sign": boolean,
      "reject": boolean,
    },
  },
}
```

#### Even more verbose process definition OUT OF SCOPE
One could argue that we are implicit doing some actions that are hidden for the developer today.
If we go fullblown BPMN one could introduce UserTasks and ServiceTasks and make every implicit action we do defined in the process. This is not in scope for this issue, but should be considered for future work

Example image of the process described abov:
image

### Acceptance Criterias

- [ ] If a user don't have access at all (ie. no read or write access) the endpoint should return 403 Forbidden allowing the frontend to show a generic "you don't have access" layout.
- [ ] Backend returns information about the current logged on users access rights to the current step

### LINKED ISSUE #207: Authorize actiontype
### Description

Add support for authorizing actions

### Additional Information

_No response_

### Tasks

_No response_

### Acceptance Criterias

_No response_

### PR DESCRIPTION
**Pull request by tjololo**
## Description
Resolves #205 
Resolves #207
Rewrite of process engine and process definition (BPMN). Process navigation is now decided based on actions defined on a process task.
This requires a small rewrite of the proces.bpmn file in apps as we are moving altinn specific properties to bpmn:extensionProperties.
This also introduces additional properties in the GET /process endpoint listing available actions for the current task

POST /process/next has an additional query parameter `action` to make migration to v8 easier `elementId` param is still available and act as a alias for `action`.
Applications leveraging `elementId` to perform gateway navigation would need to change some parts of the frontend to support v8.

## Related Issue(s)
- #205
- #207

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.App.Core/Internal/Process/ProcessEngine.cs [hunk 1]
```diff
@@ -1,138 +1,289 @@
...
+            return new ProcessChangeResult()
             {
-                processChange.ProcessMessages = new List<ProcessChangeInfo>();
-                processChange.ProcessMessages.Add(new ProcessChangeInfo() { Message = $"Instance does not have current task information!", Type = "Conflict" });
-                processChange.FailedProcessChange = true;
-                return processChange;
-            }
+                Success = false,
+                ErrorMessage = "Process is already started. Use next.",
+                ErrorType = "Conflict"
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Internal/Process/ProcessEngine.cs]: Consider creating a ProcessErrorType class holding string constants to avoid magic strings as the same type is repeated.

File: src/Altinn.App.Api/Controllers/InstancesController.cs [hunk 2]
```diff
@@ -267,12 +266,19 @@ public class InstancesController : ControllerBase
...
-                processChangeContext.DontUpdateProcessAndDispatchEvents = true;
-                processChangeContext = await _processEngine.StartProcess(processChangeContext);
+                // start process and goto next task
+                ProcessStartRequest processStartRequest = new ProcessStartRequest
+                {
+                    Instance = instanceTemplate,
+                    User = User,
+                    Dryrun = true
+                };
+                var result = await _processEngine.StartProcess(processStartRequest);
```
Comment (RonnyB71) [inline src/Altinn.App.Api/Controllers/InstancesController.cs:280]: Should the result be handled in case there is an error in there?

File: src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs [hunk 3]
```diff
@@ -95,5 +104,21 @@ public class AuthorizationClient : IAuthorization
...
+
+        /// <inheritdoc />
+        public async Task<bool> AuthorizeAction(AppIdentifier appIdentifier, InstanceIdentifier instanceIdentifier, ClaimsPrincipal user, string action, string? taskId = null)
+        {
+            _logger.LogInformation("About to authorize action {Action}", action);
+            XacmlJsonRequestRoot request = DecisionHelper.CreateDecisionRequest(appIdentifier.Org, appIdentifier.App, user, action, instanceIdentifier.InstanceOwnerPartyId, instanceIdentifier.InstanceGuid, taskId);
+            XacmlJsonResponse response = await _pdp.GetDecisionForRequest(request);
+            if (response?.Response == null)
+            {
+                _logger.LogInformation("Failed to get decision from pdp: {SerializeObject}", JsonConvert.SerializeObject(request));
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs]: Warning?

File: src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs [hunk 4]
```diff
@@ -95,5 +104,21 @@ public class AuthorizationClient : IAuthorization
 
             return result;
         }
+
+        /// <inheritdoc />
+        public async Task<bool> AuthorizeAction(AppIdentifier appIdentifier, InstanceIdentifier instanceIdentifier, ClaimsPrincipal user, string action, string? taskId = null)
+        {
+            _logger.LogInformation("About to authorize action {Action}", action);
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs]: Debug?

File: src/Altinn.App.Core/Internal/Process/ProcessEventDispatcher.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    /// <summary>
+    /// Will for each process change trigger relevant Process Elements to perform the relevant change actions.
+    ///
+    /// Each implementation 
+    /// </summary>
+    private async Task HandleProcessChanges(Instance instance, List<InstanceEvent>? events, Dictionary<string, string>? prefill)
+    {
+        if (events != null)
+        {
+            foreach (InstanceEvent processEvent in events)
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Internal/Process/ProcessEventDispatcher.cs]: InstanceEvent and ProcessEvent is used interchangeably and is a bit confusing. I see the process events is within the InstanceEventType which seems be the root of my confusion.

File: src/Altinn.App.Core/Internal/Process/ProcessEventDispatcher.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        _logger = logger;
+    }
+
+    /// <inheritdoc/>
+    public async Task<Instance> UpdateProcessAndDispatchEvents(Instance instance, Dictionary<string, string>? prefill, List<InstanceEvent>? events)
+    {
+        await HandleProcessChanges(instance, events, prefill);
+
+        // need to update the instance process and then the instance in case appbase has changed it, e.g. endEvent sets status.archived
+        Instance updatedInstance = await _instanceService.UpdateProcess(instance);
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Internal/Process/ProcessEventDispatcher.cs:48]: Should this class, which is responsible for dispatching events, update and save the process state on the instance? Seems like a side effect.

File: src/Altinn.App.Core/Internal/Process/ProcessEngine.cs [hunk 7]
```diff
@@ -1,138 +1,289 @@
...
+                AuthenticationLevel = user.GetAuthenticationLevel(),
+                OrgId = user.GetOrg()
+            },
+            ProcessInfo = instance.Process,
+        };
 
-            return await _processChangeHandler.HandleStart(processChange);
+        if (string.IsNullOrEmpty(instanceEvent.User.OrgId) && userId != null)
+        {
+            UserProfile up = await _profileService.GetUserProfile((int)userId);
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Internal/Process/ProcessEngine.cs:270]: Seems like a http call that should have been done outside of this method, ideally done once for the entire user request and just passed in as a dependency.

File: src/Altinn.App.Api/Controllers/InstancesController.cs [hunk 8]
```diff
@@ -535,22 +557,28 @@ public class InstancesController : ControllerBase
...
-            await _processEngine.StartTask(processChangeContext);
+            ProcessStartRequest rerunRequest = new()
+            {
+                Instance = targetInstance,
+                Dryrun = false,
+                User = User
+            };
+            await _processEngine.UpdateInstanceAndRerunEvents(rerunRequest, startResult.ProcessStateChange?.Events);
 
             await RegisterEvent("app.instance.created", targetInstance);
```
Comment (RonnyB71) [inline src/Altinn.App.Api/Controllers/InstancesController.cs:588]: What is the `app.instance.created` event compare to the `InstanceEventType.Created`? Should this method have been somewhere else? I'm thinking a EventDispatcherService class with a DispatchEvent method and avoid the direct use of the IEvents which ideally should have been renamed to IEventsClient (since it's a http client).

File: src/Altinn.App.Core/Interface/IProcess.cs [hunk 9]
```diff
@@ -19,7 +19,7 @@ public interface IProcess
         /// <param name="instance">the instance</param>
         /// <param name="events">process events</param>
         /// <returns></returns>
-        public Task DispatchProcessEventsToStorage(Instance instance, List<InstanceEvent> events);
+        public Task DispatchProcessEventsToStorage(Instance instance, List<InstanceEvent>? events);
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Interface/IProcess.cs:22]: Given the name and description of this interface, does the DispatchPRocessEventsToStorage method belong here? Should this be a IProcessReader and nothing more?

Comment (RonnyB71) [inline src/Altinn.App.Core/Interface/IProcess.cs]: Given the name and description of this interface, does the DispatchPRocessEventsToStorage method belong here? Should this be a IProcessReader and nothing more?

File: src/Altinn.App.Core/Internal/Process/Elements/ExtensionElements.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
+using System.Xml.Serialization;
+using Altinn.App.Core.Internal.Process.Elements.AltinnExtensionProperties;
+
+namespace Altinn.App.Core.Internal.Process.Elements
+{
+    /// <summary>
+    /// Class representing the extension elements
+    /// </summary>
+    public class ExtensionElements
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Internal/Process/Elements/ExtensionElements.cs:9]: Perhaps BpmnExtensikonElements?

Comment (RonnyB71) [reply src/Altinn.App.Core/Internal/Process/Elements/ExtensionElements.cs:9]: Ignore that one...

File: src/Altinn.App.Core/Internal/Process/IProcessEngine.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        /// </summary>
+        Task<ProcessChangeResult> Next(ProcessNextRequest request);
+
+        /// <summary>
+        /// Update Instance and rerun instance events
+        /// </summary>
+        /// <param name="startRequest"></param>
+        /// <param name="events"></param>
+        /// <returns></returns>
+        Task<Instance> UpdateInstanceAndRerunEvents(ProcessStartRequest startRequest, List<InstanceEvent>? events);
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Internal/Process/IProcessEngine.cs:27]: This seems a bit misplaced compared to the other two and the name causes me to think of side effects and why we need to re-run events.

File: src/Altinn.App.Api/Controllers/InstancesController.cs (hunk 8 above)
Comment (tjololo) [reply src/Altinn.App.Api/Controllers/InstancesController.cs:588]: The InstanceEventType.* is Enums for what events that can happen on a instance in the eyes of storage, see DispatchEventsToStorage. The string "app.instance.created" is the eventtype that is sent to the EventsComponent.
We could create an Enum for the static events dispatched to the eventcomponent but some of them have the taskId embedded in the eventID so we would still need to support dispatching events as string I think

File: src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs (hunk 3 above)
Comment (tjololo) [reply src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs]: this might even be an Error?

File: src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs (hunk 4 above)
Comment (tjololo) [reply src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs]: removing the log statement

File: src/Altinn.App.Core/Interface/IProcess.cs (hunk 9 above)
Comment (RonnyB71) [reply src/Altinn.App.Core/Interface/IProcess.cs]: Looks like the GetProcessDefinition was the first method in this, and that the class summary originates from that one. That method seems misplaced now given the other two and the fact that we have an IProcessReader interface.

File: src/Altinn.App.Core/Internal/Process/IProcessEventDispatcher.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    /// <param name="instance">The instance with updated process</param>
+    /// <param name="prefill">Prefill data</param>
+    /// <param name="events">Events that should be dispatched</param>
+    /// <returns>Instance from storage after update</returns>
+    Task<Instance> UpdateProcessAndDispatchEvents(Instance instance, Dictionary<string, string>? prefill, List<InstanceEvent>? events);
+    /// <summary>
+    /// Dispatch events for instance to the events system if AppSettings.RegisterEventsWithEventsComponent is true
+    /// </summary>
+    /// <param name="instance">The instance to dispatch events for</param>
+    Task RegisterEventWithEventsComponent(Instance instance);
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Internal/Process/IProcessEventDispatcher.cs:22]: DispatchEvent?

File: src/Altinn.App.Core/Internal/Process/IProcessNavigator.cs [hunk 13]
```diff
@@ -0,0 +1,20 @@
+using Altinn.App.Core.Internal.Process.Elements.Base;
+using Altinn.Platform.Storage.Interface.Models;
+
+namespace Altinn.App.Core.Internal.Process
+{
+    /// <summary>
+    /// Interface used to descipt the process navigator
+    /// </summary>
+    public interface IProcessNavigator
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Internal/Process/IProcessNavigator.cs:9]: Is IProcessEngine considered as a home for this method since this has process navigation methods already?

File: src/Altinn.App.Api/Controllers/InstancesController.cs (hunk 8 above)
Comment (RonnyB71) [reply src/Altinn.App.Api/Controllers/InstancesController.cs:588]: For string support we can create a static class with string constants avoiding repeating the string multiple places. Not a big thing in this case though.

"In the eyes of storge" - lets talk about that one:)

File: src/Altinn.App.Core/Features/IProcessExclusiveGateway.cs [hunk 14]
```diff
@@ -18,6 +18,7 @@ public interface IProcessExclusiveGateway
     /// </summary>
     /// <param name="outgoingFlows">Complete list of defined flows out of gateway</param>
     /// <param name="instance">Instance where process is about to move next</param>
+    /// <param name="action">Action performed by the requester</param>
     /// <returns></returns>
-    public Task<List<SequenceFlow>> FilterAsync(List<SequenceFlow> outgoingFlows, Instance instance);
+    public Task<List<SequenceFlow>> FilterAsync(List<SequenceFlow> outgoingFlows, Instance instance, string? action);
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Features/IProcessExclusiveGateway.cs:23]: Is it correct that an exclusive gateway returns a list of sequence flows?

File: src/Altinn.App.Core/Interface/IProcess.cs (hunk 9 above)
Comment (tjololo) [reply src/Altinn.App.Core/Interface/IProcess.cs]: Agreed, moving it to IProcessEventDispatcher as it feels like a more obvious home for this method

File: src/Altinn.App.Core/Features/IProcessExclusiveGateway.cs (hunk 14 above)
Comment (tjololo) [reply src/Altinn.App.Core/Features/IProcessExclusiveGateway.cs:23]: Had the same thought. We need to decide if it should be possible to return more than one flow from a gateway and then choose the default (if defined), or if a gateway always needs to return one flow

File: src/Altinn.App.Core/Internal/Process/IProcessEventDispatcher.cs (hunk 12 above)
Comment (tjololo) [reply src/Altinn.App.Core/Internal/Process/IProcessEventDispatcher.cs:22]: Rename to `DispatchEventsToEventsComponent`? As we still have events being dispatched to Storage we need to specify that this is events to the eventscomponent I think

File: src/Altinn.App.Core/Interface/IProcess.cs (hunk 9 above)
Comment (tjololo) [reply src/Altinn.App.Core/Interface/IProcess.cs]: seems like it's only usage is in IProcessEventDispatcher so we can make it a private method there simplifying the interfaces

File: src/Altinn.App.Api/Controllers/InstancesController.cs [hunk 15]
```diff
@@ -288,11 +299,16 @@ public class InstancesController : ControllerBase
 
                 // get the updated instance
                 instance = await _instanceClient.GetInstance(app, org, int.Parse(instance.InstanceOwner.PartyId), Guid.Parse(instance.Id.Split("/")[1]));
-
+
```
Comment (RonnyB71) [inline src/Altinn.App.Api/Controllers/InstancesController.cs]:
```suggestion

```

File: src/Altinn.App.Core/Internal/Process/IProcessEventDispatcher.cs (hunk 12 above)
Comment (RonnyB71) [reply src/Altinn.App.Core/Internal/Process/IProcessEventDispatcher.cs:22]: Yep, agree.

File: src/Altinn.App.Core/Models/ProcessChangeResult.cs [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+    /// <summary>
+    /// Types of errors that can occur during a process change
+    /// </summary>
+    public enum ProcessErrorType
+    {
+        /// <summary>
+        /// The process change was not allowed due to the current state of the process
+        /// </summary>
+        Conflict
```
Comment (RonnyB71) [inline src/Altinn.App.Core/Models/ProcessChangeResult.cs:35]: Is this being used?

**Review by RonnyB71 — APPROVED** [review]

File: src/Altinn.App.Core/Models/ProcessChangeResult.cs (hunk 16 above)
Comment (tjololo) [reply src/Altinn.App.Core/Models/ProcessChangeResult.cs:35]: Only possible return today is Conflict so not really used. If error results emerges we would need to use it

File: src/Altinn.App.Core/Internal/Process/IProcessEngine.cs (hunk 11 above)
Comment (tjololo) [reply src/Altinn.App.Core/Internal/Process/IProcessEngine.cs:27]: Added an issue for tracking this: #240

File: src/Altinn.App.Core/Internal/Process/IProcessNavigator.cs (hunk 13 above)
Comment (tjololo) [reply src/Altinn.App.Core/Internal/Process/IProcessNavigator.cs:9]: This is whats left of the IFlowHydration interface. I didn't want to merge it with the processengine yet as that code really still needs alot of cleanup.
Leaving it for now until the IProcessEngine is more clean
